src/gui/touchscreenlayout.cpp

```diff
@@ -0,0 +1,348 @@
+// Luanti
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2024 grorp, Gregor Parzefall <grorp@posteo.de>
+
+#include "touchscreenlayout.h"
+#include "gettext.h"
+#include "client/renderingengine.h"
+#include "client/texturesource.h"
+#include "settings.h"
+#include "convert_json.h"
+#include <json/json.h>
+
+#include "IGUIStaticText.h"
+#include "IGUIFont.h"
+
+const char *button_names[] = {
+	"jump",
+	"sneak",
+	"zoom",
+	"aux1",
+	"overflow",
+
+	"chat",
+	"inventory",
+	"drop",
+	"exit",
+
+	"fly",
+	"fast",
+	"noclip",
+	"debug",
+	"camera",
+	"range",
+	"minimap",
+	"toggle_chat",
+
+	"joystick_off",
+	"joystick_bg",
+	"joystick_center",
+};
+
+// compare with GUIKeyChangeMenu::init_keys
+const char *button_titles[] = {
+	N_("Jump"),
+	N_("Sneak"),
+	N_("Zoom"),
+	N_("Aux1"),
+	N_("Overflow menu"),
+
+	N_("Chat"),
+	N_("Inventory"),
+	N_("Drop"),
+	N_("Exit"),
+
+	N_("Toggle fly"),
+	N_("Toggle fast"),
+	N_("Toggle noclip"),
+	N_("Toggle debug"),
+	N_("Change camera"),
+	N_("Range select"),
+	N_("Toggle minimap"),
+	N_("Toggle chat log"),
+
+	N_("Joystick"),
+	N_("Joystick"),
+	N_("Joystick"),
+};
+
+const char *button_image_names[] = {
+	"jump_btn.png",
+	"down.png",
+	"zoom.png",
+	"aux1_btn.png",
+	"overflow_btn.png",
+
+	"chat_btn.png",
+	"inventory_btn.png",
+	"drop_btn.png",
+	"exit_btn.png",
+
+	"fly_btn.png",
+	"fast_btn.png",
+	"noclip_btn.png",
+	"debug_btn.png",
+	"camera_btn.png",
+	"rangeview_btn.png",
+	"minimap_btn.png",
+	// toggle button: switches between "chat_hide_btn.png" and "chat_show_btn.png"
+	"chat_hide_btn.png",
+
+	"joystick_off.png",
+	"joystick_bg.png",
+	"joystick_center.png",
+};
+
+v2s32 ButtonMeta::getPos(v2u32 screensize, s32 button_size) const
+{
+	return v2s32((position.X * screensize.X) + (offset.X * button_size),
+			(position.Y * screensize.Y) + (offset.Y * button_size));
+}
+
+void ButtonMeta::setPos(v2s32 pos, v2u32 screensize, s32 button_size)
+{
+	v2s32 third(screensize.X / 3, screensize.Y / 3);
+
+	if (pos.X < third.X)
+		position.X = 0.0f;
+	else if (pos.X < 2 * third.X)
+		position.X = 0.5f;
+	else
+		position.X = 1.0f;
+
+	if (pos.Y < third.Y)
+		position.Y = 0.0f;
+	else if (pos.Y < 2 * third.Y)
+		position.Y = 0.5f;
+	else
+		position.Y = 1.0f;
+
+	offset.X = (pos.X - (position.X * screensize.X)) / button_size;
+	offset.Y = (pos.Y - (position.Y * screensize.Y)) / button_size;
+}
+
+bool ButtonLayout::isButtonAllowed(touch_gui_button_id id)
+{
+	return id != joystick_off_id && id != joystick_bg_id && id != joystick_center_id &&
+			id != touch_gui_button_id_END;
+}
+
+bool ButtonLayout::isButtonRequired(touch_gui_button_id id)
+{
+	return id == overflow_id;
+}
+
+s32 ButtonLayout::getButtonSize(v2u32 screensize)
+{
+	return std::min(screensize.Y / 4.5f,
+			RenderingEngine::getDisplayDensity() * 65.0f *
+					g_settings->getFloat("hud_scaling"));
+}
+
+ButtonLayout ButtonLayout::getDefault()
+{
+	return {{
```

Comment:
This looks like a constant, thus please move it up (if possible). The function is not necessary. Use the constant in `loadFromSettings`.